Refactor OpenAI flavor with SDK

TomeHirata opened the pull request:
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/TomeHirata/mlflow/pull/13774?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/13774/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 13774
```

</p>
</details>

### Related Issues/PRs

N/A

### What changes are proposed in this pull request?

- Replaced the custom API call logic with OpenAI SDK: https://github.com/openai/openai-python?tab=readme-ov-file#retries

### How is this PR tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests

<!-- Attach code, screenshot, video used for manual testing here. -->
### Verification
#### Chat.Completion
<img width="1015" alt="Screenshot 2024-11-13 at 11 43 43" src="https://github.com/user-attachments/assets/504fb91d-2abd-42eb-bd4c-d0bf43b84975">

#### Legacy Completion
<img width="1057" alt="Screenshot 2024-11-13 at 13 06 32" src="https://github.com/user-attachments/assets/de842d1c-ae42-4fbb-841e-daf31fa9a0f0">

#### Embedding
<img width="1057" alt="Screenshot 2024-11-13 at 12 03 17" src="https://github.com/user-attachments/assets/ef876abf-1e9e-4342-bf5f-1cc1325c32a6">

#### Deployment
<img width="770" alt="image" src="https://github.com/user-attachments/assets/f9a13707-66ee-4c67-a0af-18acb7e0b3aa">

#### Tracing
Verified that the following code stored the model file and spans correctly.
```python
mlflow.openai.autolog(
    log_model_signatures=True,
    log_models=True,
    log_datasets=True)
mlflow.openai.autolog(
    log_model_signatures=True,
    log_models=True,
    log_datasets=True)

mlflow.set_experiment("OpenAI")

openai_client = openai.OpenAI(api_key=os.environ.get("OPENAI_API_KEY"))

messages = [
    {
        "role": "user",
        "content": "How can I improve my resting metabolic rate most effectively?",
    }
]

response = openai_client.chat.completions.create(
    model="gpt-4o",
    messages=messages,
    temperature=0.99,
)

print(response)
```

#### Evaluation
Confirmed the [LLM evaluation quickstart code](https://mlflow.org/docs/latest/llms/llm-evaluate/index.html#quickstart) worked well.

#### Azure OpenAI
<img width="568" alt="image" src="https://github.com/user-attachments/assets/119a6f50-da2e-4e89-a671-ee193ce0b594">


### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Release Notes

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

<!-- Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change. -->

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

<!-- patch -->

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

TomeHirata committed:
refactor OpenAI flavor with SDK

Signed-off-by: TomuHirata <tomu.hirata@gmail.com>

TomeHirata committed:
remove unnecessary refresh

Signed-off-by: TomuHirata <tomu.hirata@gmail.com>

TomeHirata committed:
remove request params construction

Signed-off-by: TomuHirata <tomu.hirata@gmail.com>

TomeHirata committed:
refactor openai api client

Signed-off-by: TomuHirata <tomu.hirata@gmail.com>

TomeHirata committed:
add exception case

Signed-off-by: TomuHirata <tomu.hirata@gmail.com>

harupy commented:
Is it possible to split this PR into two? One for deployments client, and one for openai flavor?

harupy commented on mlflow/openai/api_request_parallel_processor.py:
```diff
@@ -71,279 +59,55 @@ def increment_num_rate_limit_errors(self):
-    max_tokens_per_minute: float = 90_000,
-    token_encoding_name: str = "cl100k_base",
-    max_attempts: int = 5,
+    requests: list[Callable],
```
```suggestion
    requests: list[Callable[<arg type>, <return type>]],
```

Let's add argument type and return type.

TomeHirata commented:
@harupy I was not intended to change the deployment logic but needed to change together since an internal method for the flavor was used in the deployment. One workaround is defining a separate method for concurrent call and use it in the flavor first, but it may complicate the process too.

harupy commented on mlflow/deployments/openai/__init__.py:
```diff
@@ -105,54 +100,30 @@ def predict(self, deployment_name=None, inputs=None, endpoint=None):
-            request_url = REQUEST_URL_CHAT
+            from openai import OpenAI
+
+            client = OpenAI(
+                api_key=api_token.token,
+                base_url=api_config.api_base,
+            )
```
Can we instantiate the client with `max_retries` and `timeout`?

TomeHirata replied:
We could, I implemented it in this way to be aligned with how batch size is used.

harupy commented on mlflow/openai/__init__.py:
```diff
@@ -698,91 +668,103 @@ def get_params_list(self, data):
 
         first_string_column = _first_string_column(data)
         texts = data[first_string_column].tolist()
+
```
```suggestion
        client = self.client.with_options(
            max_retries=max_retries, timeout=timeout
        )
```

can we update the options here?

harupy commented on mlflow/openai/api_request_parallel_processor.py:
```diff
@@ -71,279 +59,55 @@ def increment_num_rate_limit_errors(self):
     max_workers: int = 10,
-    throw_original_error=False,
-    extra_headers: Optional[dict[str, str]] = None,
+    throw_original_error=True,
```
```suggestion
    throw_original_error: bool = True,
```

harupy commented on mlflow/openai/api_request_parallel_processor.py:
```diff
@@ -71,279 +59,55 @@ def increment_num_rate_limit_errors(self):
-            time.sleep(0.001)  # avoid busy waiting
+            for index, task in requests_iter
+        ]
+        wait(futures, return_when=ALL_COMPLETED)
```
```suggestion
        wait(futures, return_when=FIRST_EXCEPTION)
```

should we use `FIRST_EXCEPTION` to fail fast?

TomeHirata replied:
I just tried to keep the existing logic here, the current logic let all tasks run and count the number of failures. I personally agree with fail fast.

harupy commented on mlflow/deployments/openai/__init__.py:
```diff
@@ -105,54 +100,30 @@ def predict(self, deployment_name=None, inputs=None, endpoint=None):
+        if api_config.api_type == "azure":
+            from openai import AzureOpenAI
+
+            client = AzureOpenAI(
```
I'd make the change if this class doesn't have to be serializable :) This can be addressed in a follow-up.

harupy commented on mlflow/deployments/openai/__init__.py:
```diff
@@ -105,54 +100,30 @@ def predict(self, deployment_name=None, inputs=None, endpoint=None):
-            )[0]
-        except MlflowException:
-            raise
+            return client.with_options(
+                max_retries=api_config.max_retries, timeout=api_config.timeout
+            ).chat.completions.create(messages=inputs["messages"], model=endpoint)
```
can we update the options here? That's unlikely to fail.

```suggestion
         client = client.with_options(
              max_retries=api_config.max_retries, timeout=api_config.timeout
          )
         try:
            return client.chat.completions.create(messages=inputs["messages"], model=endpoint)
```

harupy commented on mlflow/openai/__init__.py:
```diff
@@ -698,91 +668,103 @@ def get_params_list(self, data):
+                model=self.model["model"],
+                **params,
+            )
+            for message in messages_list
```
```suggestion
            for messages in messages_list
```

nit

harupy commented on mlflow/openai/__init__.py:
```diff
@@ -698,91 +668,103 @@ def get_params_list(self, data):
         requests = [
-            {
-                **self.model,
+            partial(
+                self.client.with_options(
+                    max_retries=max_retries, timeout=timeout
+                ).completions.create,
```
same here: https://github.com/mlflow/mlflow/pull/13774/files#r1839926386

TomeHirata committed:
Refactor OAI evaluation metric

Signed-off-by: TomuHirata <tomu.hirata@gmail.com>

TomeHirata committed:
recover proxy_url

Signed-off-by: TomuHirata <tomu.hirata@gmail.com>

TomeHirata committed:
ci test

Signed-off-by: TomuHirata <tomu.hirata@gmail.com>

B-Step62 commented on mlflow/deployments/openai/__init__.py:
```diff
@@ -105,54 +100,31 @@ def predict(self, deployment_name=None, inputs=None, endpoint=None):
+                messages=inputs["messages"], model=endpoint
+            ).model_dump()
         except Exception as e:
             raise MlflowException(f"Error response from OpenAI:\n {e}")
```
Do we need this exception translation?

TomeHirata replied:
I thought this is for not exposing raw exceptions of external libraries?

serena-ruan replied:
This try catch seems useless and confusing (error in a string format is not easy to see), could we remove it?

TomeHirata replied:
Sure, removed!

B-Step62 commented on mlflow/metrics/genai/model_utils.py:
```diff
@@ -99,66 +98,44 @@ def _call_openai_api(openai_uri, payload, eval_parameters, extra_headers, proxy_
+        ).with_options(max_retries=api_config.max_retries, timeout=api_config.timeout)
+
+    if proxy_url is not None:
+        client.base_url = proxy_url
```
Can you test if setting `proxy_url` works for any URL, say Databricks model serving endpoint? https://docs.databricks.com/en/machine-learning/model-serving/score-foundation-models.html#query-a-chat-completion-model


Sth like this:
```
os.environ["OPENAI_API_KEY"] = <your-databricks-access-token>

eval_data = pd.DataFrame(
    {
        "inputs": ["What is MLflow?"],
        "outputs": ["MLflow is an open-source platform for managing machine learning processes. It provides tools for tracking experiments, packaging code into reproducible runs, and sharing and deploying models."],
        "ground_truth": ["MLflow is an OSS platform for managing ML lifecycle."],
    }
)

with mlflow.start_run():
    mlflow.evaluate(
        data=eval_data,
        extra_metrics=[
            answer_similarity(
                model="openai:/
databricks-meta-llama-3-1-70b-instruct",
                proxy_url="https://example.staging.cloud.databricks.com/serving-endpoints"
            )
        ]
```

TomeHirata replied:
This does not work well if the proxy API does not follow the path design of OpenAI. The method of SDK construct the endpoint by concatenating base url with "/chat/completion".
https://platform.openai.com/docs/models#model-endpoint-compatibility

B-Step62 replied:
Discussed offline: This code will be migrated again to gateway+REST soon ([ref](https://github.com/mlflow/mlflow/blob/master/mlflow/metrics/genai/model_utils.py#L58)) so we decided not to support proxy URL in non OpenAI format for now.

B-Step62 commented on mlflow/openai/__init__.py:
```diff
@@ -639,34 +637,6 @@ def __init__(self, model):
-        if self.request_configs.get("api_type") in ("azure", "azure_ad", "azuread"):
-            deployment_id = self.request_configs.get("deployment_id")
-            if self.request_configs.get("engine"):
-                # Avoid using both parameters as they serve the same purpose
```
Does this validation covered by OAI client?

B-Step62 replied:
> we don't manually construct parameters any more

We still construct them via `_get_api_config()` right?  I think the intention here is avoid passing `OPENAI_ENGINE` and `OPENAI_DEPLOYMENT_NAME` together, which can still happen.

So my question is whether OAI client can catch this conflict and raise an exception. Or is `OPENAI_ENGINE` fully deprecated and we can safely remove this validation?

B-Step62 replied:
Discussed offline: `OPENAI_ENGINE` seems to be deprecated a while ago so we don't need to support it.

TomeHirata replied:
Yes, engine param was replaced by "model" in OpenAI and by "deployment_id" in Azure.
https://github.com/openai/openai-python/discussions/742
https://learn.microsoft.com/en-us/azure/ai-services/openai/reference

B-Step62 commented on mlflow/openai/__init__.py:
```diff
@@ -698,91 +668,102 @@ def get_params_list(self, data):
-            max_tokens_per_minute=self.api_config.max_tokens_per_minute,
-        )
-        return [r["choices"][0]["message"]["content"] for r in results]
+        client = self.client.with_options(max_retries=max_retries, timeout=timeout)
```
Very nicee😄

B-Step62 commented on mlflow/openai/__init__.py:
```diff
@@ -698,91 +668,102 @@ def get_params_list(self, data):
-            {
-                **self.model,
+            partial(
+                client.completions.create,
```
Not related to this PR: OpenAI deprecated completion API a while ago, do you think it makes sense to drop support in MLflow as well? cc: @BenWilson2 @harupy

harupy replied:
I think we can remove it once OpenAI removes it: https://community.openai.com/t/completion-models-are-now-considered-legacy/656302/3 implies it's marked as legacy but not deprecated yet.

harupy replied:
I'm curious if other LLM packages still support it or not.

B-Step62 replied:
Yea aligns with OpenAI's status makes sense

BenWilson2 replied:
Most integration packages have dropped it, but we still have stuff on our runtime side that supports it (legacy endpoints). We'll wait for everything to be cleaned up and removed before dropping support :)

B-Step62 commented on mlflow/openai/api_request_parallel_processor.py:
```diff
@@ -71,283 +59,59 @@ def increment_num_rate_limit_errors(self):
     max_workers: int = 10,
-    throw_original_error=False,
-    extra_headers: Optional[dict[str, str]] = None,
+    throw_original_error: bool = True,
```
It seems we always pass `throw_original_error=True`. Can we drop this argument to simplify the error handling logic?

B-Step62 commented on mlflow/openai/api_request_parallel_processor.py:
```diff
@@ -71,283 +59,59 @@ def increment_num_rate_limit_errors(self):
-    max_tokens_per_minute: float = 90_000,
-    token_encoding_name: str = "cl100k_base",
-    max_attempts: int = 5,
+    requests: list[Callable[[], Any]],
```
nit: `requests` is a bit confusing as it actually represents a list of partial functions to execute. Can we use `tasks` instead? It will align better with `for index, task in requests_iter` below.

TomeHirata replied:
Yeah, I first used tasks and then changed to requests since I thought tasks are a bit too general. Maybe we can go with something like `api task`?

harupy replied:
how about `request_task`?

TomeHirata replied:
SGTM, let me change the naming here

B-Step62 commented on mlflow/utils/openai_utils.py:
```diff
@@ -140,6 +132,8 @@ class _OpenAIApiConfig(NamedTuple):
     api_base: str
     engine: Optional[str]
     deployment_id: Optional[str]
+    max_retries: int = 10
```
Is there a particular reason you picked 10 for retry count here? It seems our original implementation used `5` and OpenAI [default](https://github.com/openai/openai-python/blob/52357cff50bee57ef442e94d78a0de38b4173fc2/src/openai/_constants.py#L10) is `2`.

TomeHirata replied:
Sorry, this is for ci debug. Some tests in ci is flacky and I couldn't reproduce it. Definitely plan to set it to 5 to keep the existing setting.

B-Step62 commented on mlflow/metrics/genai/model_utils.py:
```diff
@@ -99,66 +98,44 @@ def _call_openai_api(openai_uri, payload, eval_parameters, extra_headers, proxy_
-        raise e
+            **eval_parameters,
+        )
+        # to_dict is not available before openai v1.17.0
```
Great catch!!

B-Step62 reviewed (commented):
Super fast work! Left a few comments but the new implementation looks much cleaner and maintainable😄

harupy commented on mlflow/deployments/openai/__init__.py:
```diff
@@ -105,54 +100,31 @@ def predict(self, deployment_name=None, inputs=None, endpoint=None):
+                api_key=api_token.token,
+                azure_endpoint=api_config.api_base,
+                api_version=api_config.api_version,
+                azure_deployment=api_config.deployment_id,
+            ).with_options(max_retries=api_config.max_retries, timeout=api_config.timeout)
```
```suggestion
                azure_deployment=api_config.deployment_id,
                max_retries=api_config.max_retries,
                timeout=api_config.timeout
            )
```

can we pass the options?

TomeHirata replied:
Sorry, what is the aim of this change? azure_deployment is already added during the instantiation of the client.

harupy replied:
Can we also pass `max_retries` and `timeout` to the constructor?

TomeHirata replied:
I see, good call, let me move the parameters to the constructor.

TomeHirata committed:
address review comments

Signed-off-by: TomuHirata <tomu.hirata@gmail.com>

TomeHirata committed:
Refactor api call method

Signed-off-by: TomuHirata <tomu.hirata@gmail.com>

harupy commented on mlflow/metrics/genai/model_utils.py:
```diff
@@ -99,66 +98,45 @@ def _call_openai_api(openai_uri, payload, eval_parameters, extra_headers, proxy_
+            **eval_parameters,
+        )
+        # to_dict is not available before openai v1.17.0
+        return _parse_chat_response_format(response.model_dump())
```
Do we need to call `model_dump`? Can we return `resp.choices[0].message.content`?

TomeHirata replied:
We could, just wanted to reuse _parse_chat_response_format if possible and this object response is harder to mimic in tests as well.

harupy replied:
I see, can you add a TODO comment here? I think we can use the mock openai server to simplify tests.

harupy replied:
Can we add a TODO comment? I think we can use the mock openai server to remove mocking.

TomeHirata committed:
fix the construction of client

Signed-off-by: TomuHirata <tomu.hirata@gmail.com>

TomeHirata committed:
add more comments

Signed-off-by: TomuHirata <tomu.hirata@gmail.com>

TomeHirata commented:
Hi, @B-Step62 @harupy. Thank you for the reviews. I've addressed most comments and fixed the flaky tests with old versions of OAI SDK, which seemed to be related to `with_option` method (https://github.com/openai/openai-python/issues/865). Can you take another look?

harupy commented on mlflow/openai/api_request_parallel_processor.py:
```diff
@@ -71,283 +59,58 @@ def increment_num_rate_limit_errors(self):
+    """Processes API requests in parallel"""
     # initialize trackers
-    retry_queue = queue.Queue()
     status_tracker = StatusTracker()  # single instance to track a collection of variables
```
Do we still need `StatusTracker`? If not, can we remove `StatusTracker` in a follow-up?

harupy commented on mlflow/openai/api_request_parallel_processor.py:
```diff
@@ -71,283 +59,58 @@ def increment_num_rate_limit_errors(self):
-            time.sleep(0.001)  # avoid busy waiting
+            for index, task in request_tasks_iter
+        ]
+        wait(futures, return_when=FIRST_EXCEPTION)
```
We can keep this line, but I found it's useless because when the executor exits, it waits for all the running and pending futures to complete.

```python
with ThreadPoolExecutor(max_workers=max_workers) as executor:

    wait(futures, return_when=FIRST_EXCEPTION)

# executor.__exit__ calls `executor.shutdown()` here and waits for both running and pending futures to copmlete.
```

Docs: https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Executor.shutdown

TomeHirata replied:
I see, then I guess `StatusTracker` is still useful to tell the summary stats of the execution?

harupy replied:
Suppose we have 1M requests, and the first request fails for some reason. In this case, we want to cancel the pending (1M - 1) requests to save money, but the current implementation waits for 1M requests to complete, then fails. I don't think we need to solve this in this PR.

harupy replied:
We might need to call `shutdown(wait=True, cancel_futures=True)` before `executor.__exit__` calls `executor.shutdown(wait=True, cancel_futures=False)` as shown below:

```python
with concurrent.futures.ThreadPoolExecutor() as executor:
    futures: list[concurrent.futures.Future] = [...]
    res = concurrent.futures.wait(futures, return_when=concurrent.futures.FIRST_EXCEPTION)
    for f in res.done:
        try:
            f.result()  # this is a no-op for successful futures
        except Exception:
            executor.shutdown(wait=True, cancel_futures=True)
            raise
```

TomeHirata replied:
Understood your point now. Then can we change the logic like this?
```
with ThreadPoolExecutor(max_workers=max_workers) as executor:
        futures = [
            executor.submit(
                call_api,
                index=index,
                task=task,
                results=results,
                status_tracker=status_tracker,
            )
            for index, task in request_tasks_iter
        ]
        wait(futures, return_when=FIRST_EXCEPTION)
        executor.shutdown(wait=True, cancel_futures=True)
```

harupy replied:
yeah let's address this problem in a follow-up. I'm not sure if this really. Let's create a simple example to make sure this behaves as expected.

TomeHirata replied:
I actually tested the behavior with the code below, but ok to defer this problem to follow up
<img width="517" alt="image" src="https://github.com/user-attachments/assets/fc3783ef-f506-4d98-ab67-f4f182984fc2">

B-Step62 replied:
> Suppose we have 1M requests, and the first request fails for some reason. In this case, we want to cancel the pending (1M - 1) requests to save cost, but the current implementation waits for 1M requests to complete, then fails.

Actually we don't necessarily want to shut down other requests. If some requests may fail transient reason or invalid payload, they should be treated as partial failure rather than throwing all items in the batch away.

harupy reviewed (approved):
Left a few more comments, otherwise LGTM!

TomeHirata committed:
add openai to dependencies of deployment test

Signed-off-by: TomuHirata <tomu.hirata@gmail.com>

B-Step62 commented on mlflow/utils/openai_utils.py:
```diff
@@ -140,6 +132,8 @@ class _OpenAIApiConfig(NamedTuple):
     api_base: str
     engine: Optional[str]
```
nit: Let's remove `engine` for clarity.

B-Step62 reviewed (approved):
LGTM! Love the huge simplification😄

BenWilson2 reviewed (approved):
Looks good! Let's verify that tracing functions as intended with this change as well :)

TomeHirata committed:
remove engine parameter

Signed-off-by: TomuHirata <tomu.hirata@gmail.com>

TomeHirata committed:
remove unnecessary error conversion

Signed-off-by: TomuHirata <tomu.hirata@gmail.com>
